### DISCUSSION TITLE
Functions and Nodes in Observable

### DISCUSSION THREAD

**Discussion by TajVirani** [LegendApp/legend-state#359] [Q&A]
I'm curious what the stance is on this with Legend-State. I know with things like Redux its not something you want to do as the methodology around Redux is that the entire state can be serialized, I was wondering if that is the same principle here.

A use case we have is that we have a monorepo with an npm workspaces folder for common modules, one being a websocket context that handles sending and receiving messages and executing functions based on specific topics. Our solution is to register with a useRef inside the provider the topic label and a function to execute. Now that this component is being moved to the workspace, I'm having a hard time figuring out how to get access to the map of labels / functions and maybe use Legend-State? Otherwise I can hoist the ref outside of the provider.... but with other modules we expose the legend-state state so that the application using the module can proxy their global state to the module state for ease of use and keeping this consistent across all modules would be ideal.

Any advice or guidance is much appreciated.

On an unrelated note, I started making a plugin like redux devtools to see global state as it updates, I didn't see anything already available. Is that a thing somewhere I just don't know about? Or is this something that is currently needed?

**Comment by jmeistrich** [comment]
Putting functions into an observable is totally fine, and a recommended pattern. Functions get extracted from the raw data and put into observable metadata, so when it's serialized into persistence it doesn't include the functions.

Does that answer your question? If not could you share a bit of code? I'm not totally following what you're trying to do.

A plugin like redux devtools is very much needed. We're planning to start building out dev tools like that after we release v3, but if you want to start it now that would be a huge help!

**Reply by TajVirani** [reply]
So I have a Websocket context that I'm loading from a common folder called modules. It handles messages like this sone { Topic: "cart.add", Data: { itemId: 'id' etc... }

So we need to execute a function in the UI of adding an item to the cart in this instance when we get the topic cart.add.

So I have a legendstate observable called "messageTopicMap"

`const messageTopicMap = obserbable<{[topic: string]: (...args: any[]) => void>;`

I have another piece of code that registers all our methods when the socket connection is made and does something like this for every topic

```
export default class ItemMessaging {

static register() {
messageTopicMap["item.addedorupdated"].set(itemMessaging.addedorupdated);
}

static itemMessaging.addedorupdated(args) { 
some things I want to do...
}
}
```
I also tried mergeIntoObservable as well.

```
const newMethods = {
      "item.added": ItemMessaging.addedorupdated,
      "item.updated": ItemMessaging.addedorupdated,
    };
    mergeIntoObservable(messageTopicMap, newMethods);
  }
```

Then in my websocket context, I have code that executes on a new message received:

```
const handler = messageTopicMap[message.Topic].peek();
        if (handler) {
          handler(message, getLimiter);
        }
```

But no matter how I write it typescript says its either {} or not a valid function. I tried  messageTopicMap[message.Topic](message, getLimiter), I tried  handler = messageTopicMap[message.Topic](). Does not want to play nice.

I'm glad we can put in functions though. I am probably not doing this correctly in the slightest.

**Reply by PwGawzynski** [reply]
You are not alone. I’m trying to do something similar: I’d like to set a function as an observable to decouple components and overcome props drilling, but I don’t know why—it gives me an error when I set the function as an observable:

```
This expression is not callable.
  Type '{}' has no call signatures.
```

 
![image](https://github.com/user-attachments/assets/678bd541-a820-4370-a81f-7621f9ad1a7c)

Probably not do it in correct way to but in doc there is not about this case :(

**Comment by apperside** [comment]
Hi,
I've just published a library that acts as an embedded dev tool and allows viewing and editing any observable in real time with two way binding (you can edit the values and see them reflected on the ui).
I didn't put yet a lot of effort on that, it's just something I built because I started working with legend state on a project and not having something to inspect the state was very painful, but still very useful and far too be something production ready. 

It's just a minimal setup, there are a lot of things that could be done, like tracking history to know what's happened and many more things, but for now I'll just wait for some feedback to go forward. Any contribution/opinion/feedback will be very appreciated.

You can find it  [here](https://github.com/apperside/legend-state-dev-tools )
